hypothesis-python/src/hypothesis/stateful.py

```diff
@@ -73,52 +70,61 @@ def __delete__(self, obj):
+def run_state_machine_as_test(state_machine_factory, settings=None):
+    """Run a state machine definition as a test, either silently doing nothing
+    or printing a minimal breaking program and raising an exception.
+
+    state_machine_factory is anything which returns an instance of
+    GenericStateMachine when called with no arguments - it can be a class or a
+    function. settings will be used to control the execution of the test.
+    """
     if settings is None:
         try:
             settings = state_machine_factory.TestCase.settings
+            assert isinstance(settings, Settings)
         except AttributeError:
             settings = Settings.default
 
-    search_strategy = StateMachineSearchStrategy(settings)
+    @settings
+    @given(st.data())
+    def run_state_machine(machine, data):
+        state_machine = machine()
+        data.data.hypothesis_runner = state_machine
 
-    return find(
-        search_strategy,
-        is_breaking_run,
-        settings=settings,
-        database_key=state_machine_factory.__name__.encode('utf-8')
-    )
+        n_steps = Settings.default.stateful_step_count
+        should_continue = cu.many(
+            data.data, min_size=1, max_size=n_steps, average_size=n_steps
+        )
 
+        print_steps = current_build_context().is_final or \
+            current_verbosity() >= Verbosity.debug
+        try:
+            if print_steps:
+                state_machine.print_start()
+            state_machine.check_invariants()
 
-def run_state_machine_as_test(state_machine_factory, settings=None):
-    """Run a state machine definition as a test, either silently doing nothing
-    or printing a minimal breaking program and raising an exception.
+            while should_continue.more():
+                value = data.data.draw(state_machine.steps())
+                if print_steps:
+                    state_machine.print_step(value)
+                state_machine.execute_step(value)
+                state_machine.check_invariants()
+        finally:
+            if print_steps:
+                state_machine.print_end()
+            state_machine.teardown()
 
-    state_machine_factory is anything which returns an instance of
-    GenericStateMachine when called with no arguments - it can be a class or a
-    function. settings will be used to control the execution of the test.
-    """
-    try:
-        breaker = find_breaking_runner(state_machine_factory, settings)
-    except NoSuchExample:
-        return
-    try:
-        with BuildContext(None, is_final=True):
-            breaker.run(state_machine_factory(), print_steps=True)
-    except StopTest:
-        pass
-    raise Flaky(
-        u'Run failed initially but succeeded on a second try'
-    )
+    # Use a machine digest to identify stateful tests in the example database
+    run_state_machine.hypothesis.inner_test._hypothesis_internal_add_digest = \
+        function_digest(state_machine_factory)
```

Comment:
Does this do the right thing when `state_machine_factory` is a class?

Reply:
Yep, `function_digest` works on classes as well as functions, even (I just checked) old-style classes on Python 2.